domain/src/main/java/com/teamagam/gimelgimel/domain/layers/ProcessNewVectorLayerInteractor.java

```diff
@@ -62,39 +62,62 @@
     protected Iterable<SubscriptionRequest> buildSubscriptionRequests(
             DataSubscriptionRequest.SubscriptionRequestFactory factory) {
         DataSubscriptionRequest dataSubscriptionRequest = factory.create(
-                buildProcessingObservable());
+                buildObservable());
         return Collections.singletonList(dataSubscriptionRequest);
     }
 
-    private Observable<URI> buildProcessingObservable() {
-        return Observable.just(null)
-                .flatMap(x -> fetchCachedURI())
-                .doOnNext(uri -> sLogger.i("VectorLayer cached uri:" + uri.toString()))
+    private Observable<URI> buildObservable() {
+        return Observable.just(mVectorLayer)
+                .flatMap(this::processIfNeeded);
+    }
+
+    private Observable<URI> processIfNeeded(VectorLayer vl) {
+        if (isOutdatedVectorLayer(vl)) {
+            sLogger.d("Not processing following vector layer because it's outdated: " + vl);
+            return Observable.empty();
+        }
+        sLogger.d("Processing vector layer " + vl);
+        return buildProcessObservable(vl);
+    }
+
+    private boolean isOutdatedVectorLayer(VectorLayer vl) {
+        return mVectorLayerRepository.contains(vl.getId()) &&
+                mVectorLayerRepository.get(vl.getId()).getVersion() > vl.getVersion();
+    }
+
+    private Observable<URI> buildProcessObservable(VectorLayer vectorLayer) {
+        return Observable.just(vectorLayer)
+                .flatMap(this::cacheLayer)
+                .doOnNext(uri -> sLogger.d("Vector layer " + vectorLayer + " is cached at " + uri))
                 .doOnNext(uri -> addToRepository())
                 .doOnNext(uri -> setVisible())
                 .retryWhen(new RetryWithDelay(Constants.LAYER_CACHING_RETRIES,
                         Constants.LAYER_CACHING_RETRIES_DELAY_MS))
-                .doOnError(throwable -> sLogger.w("Couldn't cache layer " + mVectorLayer.getName()))
+                .doOnError(this::logFailure)
```

Comment:
I don't like the way that the vectorLayer can be retrieved in two different ways- as a parameter or as a class member. It's like two names for the same object and it's confusing. That the reason I used the Observable.just(null), to emphasis that the vectorLayer is a member and is not needed as a parameter.
This line explains the issue well: where does logFailer take its vectorLayer from? I expect it to receive it as a parameter...